// azle/src/lib/candid/types/constructed/variant.ts
import { CandidType } from 'azle/src/lib/candid/candid_type';
import { TypeMapping } from 'azle/src/lib/candid/type_mapping';
import { toIdlMap, CandidMap } from 'azle/src/lib/candid/types/constructed/to_idl_map';
import { IDL } from '@dfinity/candid';
import { decode } from 'azle/src/lib/candid/serde/decode';
import { encode } from 'azle/src/lib/candid/serde/encode';

export type Variant<
    T extends {
        [K in keyof T]: any;
    }
> = RequireExactlyOne<{
    [K in keyof T]: TypeMapping<T[K]>;
}>;

export function Variant<
    T extends {
        [K in keyof T]: CandidType;
    }
>(obj: T) {
    return {
        ...obj,
        tsType: {} as Variant<T>,
        toBytes(data: any) {
            return encode(this, data);
        },
        fromBytes(bytes: Uint8Array) {
            return decode(this, bytes);
        },
        getIdl(parents: any) {
            return IDL.Variant(toIdlMap(obj as CandidMap, parents));
        }
    };
}

export type RequireExactlyOne<
    ObjectType,
    KeysType extends keyof ObjectType = keyof ObjectType
> = {
    [Key in KeysType]: Required<Pick<ObjectType, Key>> &
        Partial<Record<Exclude<KeysType, Key>, never>>;
}[KeysType] &
    Omit<ObjectType, KeysType>;
