// azle/src/lib/candid/types/reference/func.ts
import { CandidType, Parent, toIdlArray } from 'azle/src/lib/candid';
import { IDL } from '@dfinity/candid';
import { Principal } from 'azle/src/lib/candid/types/reference/principal';
import { encode } from 'azle/src/lib/candid/serde/encode';
import { decode } from 'azle/src/lib/candid/serde/decode';
import { Serializable } from 'azle/src/lib/stable_structures/stable_b_tree_map';

export type Func = [Principal, string];
type Mode = 'query' | 'update' | 'oneway';

const modeToCandid = {
    query: ['query'],
    oneway: ['oneway'],
    update: []
};

export function Func(
    paramCandidTypes: CandidType[],
    returnCandidTypes: CandidType,
    mode: Mode
) {
    return {
        tsType: {} as Func,
        toBytes(data: any) {
            return encode(this, data);
        },
        fromBytes(bytes: Uint8Array) {
            return decode(this, bytes);
        },
        getIdl(parents: Parent[]) {
            return IDL.Func(
                toIdlArray(paramCandidTypes, parents),
                toIdlArray(returnCandidTypes, parents),
                modeToCandid[mode]
            );
        }
    };
}
